### PR TITLE
[2.10] more powerful proof API

### PR DESCRIPTION
**Pull request by samoht**
Expose `Tree.produce_proof` and `Tree.verify_proof` which takes a dynamic computation as a callback instead of a static set of paths as `Tree.Proof.of_keys` was exposing.

This allows generating proof for all operations supported in the Tree API (including paginated lists and folds).

On top of #1624

### CODE REVIEW

**Comment by samoht** [comment]
note: this needs more positive and negative tests.

File: src/irmin/tree.ml [hunk 1]
```diff
@@ -1935,13 +1942,38 @@ module Make (P : Private.S) = struct
...
+    let proof = Proof.of_tree tree in
+    let after = hash tree_after in
+    Proof.v ~before:kinded_hash ~after ~proof
+
+  let verify_proof p f =
+    let before = Proof.before p in
+    let after = Proof.after p in
+    let tree = Proof.to_tree (Proof.proof p) in
+    (* first check that [before] corresponds to [tree]'s hash. *)
+    if not (equal_kinded_hash before (hash tree)) then
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:1963]: Does this matter?
```diff
- if not (equal_kinded_hash before (hash tree)) then
+ if not (equal_kinded_hash before (hash ~cache:false tree)) then
```

File: src/irmin/tree.ml [hunk 2]
```diff
@@ -1760,8 +1760,12 @@ module Make (P : Private.S) = struct
 
   module Proof = struct
     type tree = t
+
+    include Proof.Make (P.Hash) (Path) (Metadata)
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:1764]: `Proof.Make.t` and `Proof.Make.proof` are different. This is a bit hard to follow 😁

Comment (samoht) [reply src/irmin/tree.ml:1764]: `t` is `(before_hash * proof * after_hash)`. I can make this clearer.

Comment (samoht) [reply src/irmin/tree.ml:1764]: I've added some docs in e695312

Comment (Ngoguey42) [reply src/irmin/tree.ml:1764]: Thanks a lot. I'll continue reviewing

File: src/irmin/tree.ml [hunk 3]
```diff
@@ -1940,13 +1947,38 @@ module Make (P : Private.S) = struct
...
-         stored in [v]. *)
-      clear tree;
-      let+ () = Lwt_list.iter_s (fun k -> find_tree tree k >|= ignore) keys in
-      of_tree tree
   end
+
+  let produce_proof repo kinded_hash f =
+    let tree = import_no_check repo kinded_hash in
+    let+ tree_after = f tree in
+    (* Here, we build a proof from [tree] (on not from [tree_after]!)
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:1955]: I think this is not working.

Consider that initial tree:
```
/foo/age = 0
/foo/version = 1
/bar/age = 2
/bar/version = 3
```

On which we apply that `f` function:
```
let f t0 =
	let increment = function None -> assert false | Some i -> Some (i + 1) in
	let* t1 = Store.Tree.update t0 [ "foo"; "age" ] increment in
	let* t2 = Store.Tree.update t0 [ "bar"; "age" ] increment in
	t2
```

#### Before creation of t1
```
t0-a:
{"Blinded_node":"af0505a618875c533867c2bfde752fe243df4cf2"}
```


#### Before creation of t2
```
t0-b:
    {"Node":[ ["bar",{                     "Blinded_node":"7214cba30e814f6ff5125646da1fef8d4dd41b82"}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["b6589fc6ab0dc82cf12099d1c2d40ab994e8410c",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}

t1-b:
    {"Node":[ ["bar",{                     "Blinded_node":"7214cba30e814f6ff5125646da1fef8d4dd41b82"}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}
```

#### After creation of t2
```
t0-c (= t0-b)
    {"Node":[ ["bar",{                     "Blinded_node":"7214cba30e814f6ff5125646da1fef8d4dd41b82"}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["b6589fc6ab0dc82cf12099d1c2d40ab994e8410c",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}

t1-c:
    {"Node":[ ["bar",{"Node":[ ["age",{    "Blinded_contents":["da4b9237bacccdf19c0760cab7aec4a8359010b0",{}]}],
                               ["version",{"Blinded_contents":["77de68daecd823babbb58edb1c8e14d7106e83bb",{}]}]]}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}

t2-c:
    {"Node":[ ["bar",{"Node":[ ["age",{    "Blinded_contents":["77de68daecd823babbb58edb1c8e14d7106e83bb",{}]}],
                               ["version",{"Blinded_contents":["77de68daecd823babbb58edb1c8e14d7106e83bb",{}]}]]}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}
```

#### Resulting proof
```
p (= t0-c) (= t0-b):
    {"Node":[ ["bar",{                     "Blinded_node":"7214cba30e814f6ff5125646da1fef8d4dd41b82"}],
              ["foo",{"Node":[ ["age",{    "Blinded_contents":["b6589fc6ab0dc82cf12099d1c2d40ab994e8410c",{}]}],
                               ["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}

{"before":{"Node":"af0505a618875c533867c2bfde752fe243df4cf2"},"after":{"Node":"f99c9fe814cb981d510563fc4807d4ec0aea2230"},"proof": ... }

```

The code: https://github.com/samoht/irmin/compare/refactor-proofs-3...Ngoguey42:problematic-produce-proof

Comment (Ngoguey42) [reply src/irmin/tree.ml:1955]: The loading of `bar` for `t2` is not reflected in `t0`

Comment (samoht) [reply src/irmin/tree.ml:1955]: Hum I'm not sure why the cache of `bar` is not shared between `t1-c` and `t0-c`. I'll investigate.

File: src/irmin/store_intf.ml [hunk 4]
```diff
@@ -371,6 +372,25 @@ module type S = sig
     val shallow : Repo.t -> kinded_hash -> tree
     (** [shallow r h] is the shallow tree object with the hash [h]. No check is
         performed to verify if [h] actually exists in [r]. *)
+
+    (** Proofs *)
```
Comment (hanshoglund) [inline src/irmin/store_intf.ml:376]:
```suggestion
    (** {1 Proofs} *)
```

**Comment by samoht** [comment]
@Ngoguey42 so actually the fix was a little bit involved as it requires adding a global cache to keep track of reading effects. The other part (adding a `Contents` case to the proof) is straightforward.

 Something like this works now:

```ocaml
  (*  The tree contains:
        [
          ([ "foo"; "age" ], "0");
          ([ "foo"; "version" ], "1");
          ([ "bar"; "age" ], "2");
          ([ "bar"; "version" ], "3");
        ]
  *)
   let f t0 =
      let* t1 = S.Tree.update t0 [ "foo"; "age" ] increment in
      let* t2 = S.Tree.update t1 [ "bar"; "age" ] increment in
      let* t3 = S.Tree.get_tree t2 [ "bar" ] in
      let* t4 = S.Tree.add_tree t2 [ "hello"; "there" ] t3 in
      let* _ = S.Tree.get t4 [ "hello"; "there"; "version" ] in
      S.Tree.remove t4 [ "bar" ]
  in
   let* p = S.Tree.produce_proof repo hash f in
   let+ _ = S.Tree.verify_proof p f in
  ()
```

e.g. so it is robust to renaming operations too.

**Comment by Ngoguey42** [comment]
The final proof of my above example is
```
    {"Node":[["bar",{"Node":[["age",{"Contents":["2",{}]}],["version",{"Blinded_contents":["77de68daecd823babbb58edb1c8e14d7106e83bb",{}]}]]}],
             ["foo",{"Node":[["age",{"Contents":["0",{}]}],["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}}
```

isn't the following intended instead? (with incremented `age`s)
```
    {"Node":[["bar",{"Node":[["age",{"Contents":["3",{}]}],["version",{"Blinded_contents":["77de68daecd823babbb58edb1c8e14d7106e83bb",{}]}]]}],
             ["foo",{"Node":[["age",{"Contents":["1",{}]}],["version",{"Blinded_contents":["356a192b7913b04c54574d18c28d46e6395428ab",{}]}]]}]]}}
```

**Comment by samoht** [comment]
No we want the minimal initial state to reproduce the computation

**Comment by Ngoguey42** [comment]
Got it!

File: src/irmin/tree_intf.ml [hunk 5]
```diff
@@ -326,20 +326,23 @@ module type S = sig
...
 
-    val of_tree : tree -> t
-    (** [of_tree t] is the proof representing the tree [t]. Shallow hashes will
-        be blinded. *)
+    type tree
 
-    val to_tree : t -> tree
-    (** [of_proof p] is the tree representing the proof [p]. Blinded parts of
-        the proof will raise [Dangling_hash] when traversed. *)
+    val of_f : tree -> (tree -> tree Lwt.t) -> proof Lwt.t
```
Comment (Ngoguey42) [inline src/irmin/tree_intf.ml:339]: Is that `of_f` really necessary? It will leak the cache out of `f` through the first parameter

File: src/irmin/store_intf.ml [hunk 6]
```diff
@@ -371,6 +372,25 @@ module type S = sig
     val shallow : Repo.t -> kinded_hash -> tree
     (** [shallow r h] is the shallow tree object with the hash [h]. No check is
         performed to verify if [h] actually exists in [r]. *)
+
+    (** {1 Proofs} *)
+
+    val produce_proof :
+      repo -> kinded_hash -> (tree -> tree Lwt.t) -> Proof.t Lwt.t
```
Comment (Ngoguey42) [inline src/irmin/store_intf.ml:379]: Let's add to the documentation that the user shouldn't hold on to the trees passed to `f`, nor to the trees derived from it. The risk being memory leaks

**Review by Ngoguey42 — COMMENTED** [review]
The way you solved the problem is smart. I'm just a bit sad that `tree.ml` is getting more and more complicated over time, forward porting this to main is going to be a challenge, but I admit that I don't have a better simple solution.

File: src/irmin/tree_intf.ml (hunk 5 above)
Comment (samoht) [reply src/irmin/tree_intf.ml:339]: yea I am not super convinced as well but otherwise, it was hard to port the tests :-) I could probably just change the tests instead.

File: src/irmin/store_intf.ml (hunk 6 above)
Comment (samoht) [reply src/irmin/store_intf.ml:379]: I think we can also try to explicitly reset the global cache at the end of the computation. That'a bit tricky with nested proofs but that could work. I'm also fine to just state this in the doc.

**Comment by samoht** [comment]
> I'm just a bit sad that tree.ml is getting more and more complicated over time,

And you yet haven't seen for my next PR which will be adding streaming effects :p

**Comment by hanshoglund** [comment]
I am doing some testing on top of this branch. Will get back with some proper feedback next week.

**Comment by samoht** [comment]
@hanshoglund I'm just pushing a few important fixes, please use those :-) (I have a very bad connection, should be up soon)

File: src/irmin/tree_intf.ml (hunk 5 above)
Comment (samoht) [reply src/irmin/tree_intf.ml]: Removed now

File: src/irmin/store_intf.ml (hunk 6 above)
Comment (samoht) [reply src/irmin/store_intf.ml:382]: Added a comment about this

File: src/irmin/tree.ml [hunk 7]
```diff
@@ -1888,7 +2047,7 @@ module Make (P : Private.S) = struct
       in
       if List.compare_length_with elts len = 0 then
         (* We have a complete proof, let's build a tree node directly! *)
-        k (Node.of_map (StepMap.of_seq (List.to_seq elts)))
+        k (Node.of_map ~env (StepMap.of_seq (List.to_seq elts)))
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:2050]: Could we drop the `complete` branch? The `else` branch could handle the `complete` case too

File: src/irmin/tree_intf.ml [hunk 8]
```diff
@@ -326,20 +326,19 @@ module type S = sig
   (** {1 Proofs} *)
 
   module Proof : sig
```
Comment (Ngoguey42) [inline src/irmin/tree_intf.ml:328]: Is that `Proof` api really necessary? Aren't `produce_proof` and `verify_proof` enough?

Comment (samoht) [reply src/irmin/tree_intf.ml:328]: It's useful in the tests so I'd like to keep them (from now)

**Comment by samoht** [comment]
I've simplified the semantics to always attach a new environment for every call to `produce_proof` and carry this along for subtrees of that proofs. We do not merge environments anymore, which means we can reset them aftet the proof is produced. It's also easier to reason about these environments as they are not over-generalised (the only remaining operation which merges environment is `Tree.merge` where it makes sense to do so).

I'll add a few negative tests and then I think it's good to merge.

**Comment by samoht** [comment]
And now with a few simple negative tests and a simplified `Proof.to_tree` signature. I think it's ready to be merged once the CI is happy. Will push further improvements in subsequent PRs if needed.

File: src/irmin/tree.ml [hunk 9]
```diff
@@ -1872,40 +2030,40 @@ module Make (P : Private.S) = struct
         match proofs with
         | [] -> k acc
         | (s, p) :: rest ->
-            tree_of_proof p (fun p -> rev_elts ((s, p) :: acc) rest k)
+            tree_of_proof ~env p (fun p -> rev_elts ((s, p) :: acc) rest k)
       in
       rev_elts [] rev_proof_steps @@ fun elts ->
       let k n =
         List.iter (fun (s, elt) -> Node.add_to_findv_cache n s elt) elts;
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:2037]: What is this addition to `findv` cache aiming? Removing doesn't seem to break the tests
https://github.com/mirage/irmin/commit/6a6f2023cfdf4bc9cc77f876949271e540f44893

Comment (Ngoguey42) [reply src/irmin/tree.ml:2037]: It does break the unit tests.

File: src/irmin/tree.ml [hunk 10]
```diff
@@ -198,9 +204,90 @@ module Make (P : Private.S) = struct
...
+        Hashes.iter
+          (fun h v ->
+            match Hashes.find_opt y h with
+            | None -> Hashes.add y h v
+            | Some v' -> assert (v == v'))
+          x
+
+    (* x' = y' <- x union y *)
+    let merge (x : t) (y : t) =
+      Fmt.epr "XXXX Env.merge\n%!";
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:269]:
```suggestion
```

**Review by Ngoguey42 — APPROVED** [review]

File: src/irmin/tree.ml (hunk 9 above)
Comment (samoht) [reply src/irmin/tree.ml:2039]: Yes, this is indeed required because otherwise we just stop at the node boundary - by filling the cache we can also connect the nodes together.

**Comment by hanshoglund** [comment]
@samoht Sorry for the delay in getting back on this. This API looks very useful indeed.

I have implemented some more extensive tests of the proof API on [this branch](https://gitlab.com/hans.hoglund/irmin/-/tree/hans@merkle-proof-experiments2). Not merge-worthy yet, but it it is of interest I can clean them up and make a PR.

**Comment by samoht** [comment]
These tests look very useful, a PR is indeed very welcome!
